DIPs/1NNN-MHH.md

```diff
@@ -0,0 +1,242 @@
+Thanks to the simplicity of textual preprocessing, C++ can enable this pattern using a macro. However, this is extremely error-prone so a common way of achieving everything mentioned so far in this article is to have a separate compilation step to collect and use information from the source code.
+
+For example, Epic Games' Unreal Engine uses the following syntax to interface between C++ and "Blueprints" 
+```C++
+class AGameActor : public AActor
+{
+  GENERATED_BODY()
+  public:
+  UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = Properties)
+  FString Name;
+  UFUNCTION(BlueprintCallable, Category = Properties)
+  FString ToString();
+};
+```
+The game then inserts hijacks the compile to read the (in our terms) user-defined attributes from the header file, however this not only hurts compilation times but is also fairly antithetical to the "D way" of doing things.
+## Description
+This DIP simply proposes adding to the language a new *SpecialKeyword* `__ATTRIBUTE__` to the grammar. 
+```diff
+SpecialKeyword:
+    __FILE__
+    __FILE_FULL_PATH__
+    __MODULE__
+    __LINE__
+    __FUNCTION__
+    __PRETTY_FUNCTION__
++   __ATTRIBUTE__
+```
+It will always resolve to a string array literal.
+```D
+void main(string[] args)
+{
+    const attr = __ATTRIBUTE__;
+    pragma(msg, typeof(attr), attr); //"const(string[]) and []"
+}
+```
+This literal shall be empty unless `__ATTRIBUTE__` is used as a default initializer, in which case it shall be resolved to either to be either empty or an array literal of the fully qualified names of all declarations a UDA is attached to, if and only if (subject to existing default initializer resolution) it is resolved to an expression within said *UserDefinedAttribute*
+
+### Some specific examples:
+A templated struct
+```D
+module testmodule;
+struct Test(string name, string[] attr = __ATTRIBUTE__)
+{
+    pragma(msg, name, " says: ", attr);
+    this(int l) {/*Do work*/}  
+}
+@Test!"name"(1) 
+int echo(int x)
+{
+    return x;
+}
+//"name says: ["testmodule.echo"]"
+```
+A simple function
+```D
+module testmodule;
+auto just(string[] at = __ATTRIBUTE__)
+{
+    return at;
+}
+
+@(just())
+int x, y, z;
+
+pragma(msg, pragma(msg, __traits(getAttributes, x)[0]));
+//"["testmodule.x", "testmodule.y", "testmodule.z"]"
+```
+A class 
+```D
+class wow {
+    string[] cont;
+    this(string[] attr = __ATTRIBUTE__)
+    {
+        cont = attr;
+    }
+}
+@(new wow)
+int cheese;
+pragma(msg, __traits(getAttributes, cheese)[0]);
+//"wow(["testmodule.cheese"])"
+```
+Finally, eliminating a `mixin`.
+```D
+template runThisFunction(string name, string[] attrs = __ATTRIBUTE__)
+{
+    pragma(msg, "DRT", attrs);
+    enum runThisFunction;
+
+    shared static this()
+    {
+        import std.stdio;
+        static foreach(at; attrs) {
+            mixin("alias theFunc = " ~ at ~ ";");
+            writef!"%s is running %s"(name, at);
+            theFunc();
+        }  
```

Comment:
You need double braces, otherwise you'd declare the symbol `theFunc` multiple times in a function which won't work.